python_interface/design/03_pyHorace_description.md

```diff
@@ -0,0 +1,269 @@
+That is, when the user does:
+
+```
+from pyHorace import Matlab
+Matlab.<func_name>(*args)
+```
+
+The `Matlab.<func_name>` part causes Python to call `Matlab.__getattr__('func_name')`. 
+This returns a Python function which wraps `horace.call_method('func_name', [], *args)`.
+Both these steps are chained by Python so that uses can do `Matlab.cut_sqw('sqw_file', [], [], [], [-1,1])` transparently.
+Because the wrapper to `call_method` is through the `__getattr__` overload of a class, users always have to use the "dot" method-access notation.
+I.e., `from pyHorace.Matlab import *; cut_sqw('sqw_file', [], [], [], [-1,1])` will not work because `pyHorace.Matlab` is not a Python module but a class.
+In addition `help(pyHorace.Matlab.cut_sqw)` will also not work because this actually refers to a (dynamically constructed) Python wrapper function.
+Thus if we want the above two functionalities (accessing Matlab "included" functions/classes and help)
+we will need to write explicit Python wrappers for *every* Matlab function we want to expose this way.
+
+The other major feature of `pyHorace` is a Python [`MatlabProxyObject`](https://github.com/mducle/hugo/blob/master/pyHorace/MatlabProxyObject.py) class
+which wraps a Matlab class so that it appears transparent to Python users.
+That is, Python users can call the methods and access the properties of any Matlab class in Python using the "dot" notation as if they are in Matlab.
+This works by overloading the special Python methods `__getattr__` and `__setattr__`
+to call the equivalent Matlab special functions `subsref` and `subassgn`
+to access the methods/properties of a class.
+In addition, the Python `__repr__`, `__str__`, `__dir__` and `__doc__` methods are also overloaded
+to provide a description of the object, its methods and help text (except that the Matlab help system is not accessible in a compiled application).
+
+This wrapper is needed because any new-style (defined by `classdef`) Matlab object returned by `call_method` and its wrappers
+is opaque to Python and cannot be handled by purely Python functions.
+Old-style Matlab objects are treated as plain `struct`s and are automatically converted to Python `dict`s,
+losing any connection with the methods defined in their `@` folders.
+To overcome this, `pyHorace` includes a [`thinwrapper`](https://github.com/mducle/hugo/blob/master/src/thinwrapper.m) class which wraps an old style class.
+Whenever an old-style class (or [struct with an old style member](https://github.com/mducle/hugo/blob/master/src/has_thin_members.m))
+is returned by a Matlab function to Python, [`call_method`](https://github.com/mducle/hugo/blob/master/src/call_method.m#L115)
+instead [wraps](https://github.com/mducle/hugo/blob/master/src/wrap_obj.m) it (recursively) by storing the actual old-style object in the Matlab base namespace
```

Comment:
Yes, you're right that it should be "successively", since old-style members of an old-style class don't need to be wrapped, just old-style fields of a struct (this was the issue with `multifit`).